README.md
# dhc4rnn
work in progress for a gradient-free optimization algorithm


ChangeLog
2016-05-24  Deniz Yuret  <dyuret@ku.edu.tr>

	* weval100: experiments low dimensional subspaces.  run gd3 in a
	subspace and report dimensionality vs best loss.

	ndims	loss@4096
	0	2.3026
	10	2.2129
	100	1.1095
	1000	0.2309
	10000	0.0207
	50816	0.0025

2016-05-23  Deniz Yuret  <dyuret@ku.edu.tr>

	* lowdim: Can we optimize on a low dimensional subspace, and how
	close would this get us to the true optimum?  GD converges in 1000
	steps, so it only explores at most a 1000-D subspace of 50K dims.
	If we knew the direction of the solution we can trivially find it
	exploring a 1-D subspace, so maybe this is not too surprizing.  If
	we picked any random 1K subspace, could we reach values close
	enough to the optimum?

2016-05-16  Deniz Yuret  <dyuret@ku.edu.tr>

	* enewton1: First try at trying to predict the newton direction.
	Step sizes determined by derivative estimates.  Slightly worse
	than ego2.  u2 (the second derivative) never goes below 0.001
	after the initial phase.  For x5 we have:

	julia> deriv12(weval, x5gpu, d5gpu)
	(-0.022166235f0,0.00053873187f0)

	So u2=0.001 is ok.  But it cannot find directions with large u1.
	Maybe play with noise?  Does not have a significant effect.
	Instead of random search for a good direction, we need to estimate
	the newton direction and make corrections to it.

	* deltaf: It takes ~10K steps to find a direction that is as good as d5 (deltaf(d5)=-0.91).
	julia> ego2c(deltaf, copy!(r,randn!(r5gpu)); ftol=-10,gmin=1e-4,step=1e7)
	(2,:f0,-0.0020427369611576126,:gnorm,1.7468734f-6,:dnorm,1000.15985f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9036666666666666)
	(3,:f0,-0.0020427369611576126,:gnorm,2.5476036f-6,:dnorm,998.5613f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9036666666666666)
	(4,:f0,-0.0020427369611576126,:gnorm,3.1720492f-6,:dnorm,1004.88354f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9036666666666666)
	(8,:f0,-0.0020427369611576126,:gnorm,4.9179957f-6,:dnorm,997.8066f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9036666666666666)
	(16,:f0,-0.0020427369611576126,:gnorm,7.0516876f-6,:dnorm,1007.56714f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9036666666666666)
	(32,:f0,-0.0021717019545803152,:gnorm,7.282012f-6,:dnorm,1005.818f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.89935)
	(64,:f0,-0.004823810417325078,:gnorm,7.250463f-6,:dnorm,1002.2937f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.8963666666666666)
	(128,:f0,-0.00628838902320593,:gnorm,6.4674446f-6,:dnorm,1005.1796f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9036333333333333)
	(256,:f0,-0.008144910173021429,:gnorm,5.4510324f-6,:dnorm,1005.2975f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9420166666666666)
	(512,:f0,-0.013292150405336638,:gnorm,5.0444105f-6,:dnorm,1004.97f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9365166666666667)
	(1024,:f0,-0.01708456265818466,:gnorm,9.710732f-6,:dnorm,1005.3696f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9134333333333333)
	(2048,:f0,-0.02366285245982498,:gnorm,9.837973f-6,:dnorm,1004.17596f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9015333333333333)
	(4096,:f0,-0.041724250299936,:gnorm,2.3839202f-5,:dnorm,1023.16656f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9256333333333333)
	(8192,:f0,-0.13053577334307326,:gnorm,6.753373f-5,:dnorm,1231.9889f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9266166666666666)
	(16384,:f0,-3.2167134764776835,:gnorm,0.0005645175f0,:dnorm,8030.281f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9144333333333333)
	(32768,:f0,-3.2167134764776835,:gnorm,0.00056755845f0,:dnorm,7961.5957f0,:step,1.0e7,:noise,1.0,:lr,1.0,:err,0.9144333333333333)


2016-05-15  Deniz Yuret  <dyuret@ku.edu.tr>

	* x1: Estimate for w*-w0:

	julia> x1 = xbuf[:,1]
	julia> weval(x1,g1)
	2.302773f0
	julia> vecnorm(g1)
	0.036010094f0
	julia> x9 = xbuf[:,9]
	julia> weval(x9,g9)
	0.049970314f0
	julia> vecnorm(g9)
	0.007846604f0
	julia> d1 = x9-x1
	julia> vecnorm(d1)
	16.732836f0
	julia> dot(d1,g1)/vecnorm(d1)
	-0.0100226f0
	julia> fd1(s)=weval(x1+s*d1/vecnorm(d1))

	First derivative accurate for exp(-7)<=eps<=exp(-1)
	Second derivative stable for exp(-4)<=eps<=exp(-1)
	julia> for i=-10:0; println((i,chebtest(fd1,5,exp(i)))); end
	(-10,(:err,0.0,-0.009337225166160881,-79.03681522581496))
	(-9,(:err,0.0,-0.009677198798914655,1.8661685640305024e-7))
	(-8,(:err,0.0,-0.010394687777943964,-11.369674966164652))
	(-7,(:err,0.0,-0.00990745106386226,1.342805505785356))
	(-6,(:err,0.0,-0.01009565346623058,-0.12870116561304296))
	(-5,(:err,0.0,-0.009972297539863784,-0.09374205470528647))
	(-4,(:err,0.0,-0.009996139184234433,-0.08263490760529549))
	(-3,(:err,0.0,-0.010029600927209776,-0.08036643844756586))
	(-2,(:err,0.0,-0.010010662646293956,-0.08068749411888569))
	(-1,(:err,0.0,-0.010059267413399381,-0.07845816156541043))
	(0,(:err,0.0,-0.012444054645358096,-0.06772726068661292))

	First derivative accurate for n>=2
	Second derivative accurate for n>=3
	julia> for n=1:10; println((n,chebtest(fd1,n,exp(-3)))); end
	(1,(:err,0.0,0.0,0.0))
	(2,(:err,-5.0067901611328125e-5,-0.010043375260390865,0.0))
	(3,(:err,-4.440892098500626e-16,-0.010050029323845916,-0.08041059287975376))
	(4,(:err,-1.9072202661618576e-7,-0.010016222311874812,-0.08039137661118922))
	(5,(:err,0.0,-0.010029600927209776,-0.08036643844756586))
	(6,(:err,-7.92827337114943e-8,-0.01003729794337197,-0.08092388592825366))
	(7,(:err,8.881784197001252e-16,-0.010026329706401443,-0.08060829138228776))
	(8,(:err,1.6959224158341613e-7,-0.010025539777791907,-0.08207477015798442))
	(9,(:err,8.881784197001252e-16,-0.010015523203612766,-0.08015735457959178))
	(10,(:err,-2.41502194242571e-7,-0.010003393668311,-0.07826335894940661))

	For n=5,eps=exp(-3) we have:
	f'=-0.010029600927209776
	f''=-0.08036643844756586
	f'/f''=0.12479837505494873
	f'^2/f''=-0.0012516778981653872

	No,no,no: second derivative is negative!!!  So we want to go an infinite step size!

	So first we want to find the direction in which the second
	derivative is most negative.  If second derivative is always
	positive, we want the direction with maximum (f')^2/f''.


model02.jl
using Knet
using Knet: Net, params, axpy!
if !isdefined(:MNIST)
    include(Pkg.dir("Knet/examples/mnist.jl"))
    mnist = minibatch(MNIST.xtrn, MNIST.ytrn, 100)
end

@knet function model02(x; winit=Gaussian(0,.1), hidden=32, f1=:relu, o...)
    h    = wf(x; out=hidden, f=f1, winit=winit)
    return wf(h; out=10, f=:soft, winit=winit)
end

# We'll try to create the two weight matrices from a single array

function initmodel{T}(w::CudaArray{T}; X=784, H=32, Y=10, B=100)
    @assert length(w) == X*H+H*Y
    f = compile(:model02; hidden=H)
    forw(f, zeros(T, X, B))
    f.reg[2].out0 = CudaArray(w.ptr, (H,X), w.dev)
    f.reg[5].out0 = CudaArray(w.ptr+H*X*sizeof(T), (Y,H), w.dev)
    return f
end

# Given a minibatch (x,y), eval model and return softloss
# Note that each minibatch gives a different loss function

evalmodel(f, x, y)=softloss(forw(f,x),y)

# We can find the right epsilon by finding the max radius at which 50%
# of the steps are positive.  For mnist, this radius seems to be about
# 0.01-0.05 starting at 0.

function deneme1(r; N=100)
    global w = fill!(CudaArray(Float32,794*32),0)
    global f = initmodel(w)
    (x,y) = first(mnist)
    @show f0 = softloss(forw(f,x),y)
    w0 = copy(w)
    global dw = copy(w)
    better=0
    for i=1:N
        scale!(r, randn!(dw))
        axpy!(1, dw, w)
        f1 = softloss(forw(f,x),y)
        f1 < f0 && (better += 1)
        copy!(w, w0)
    end
    better/N
end

# Simple algorithm:
# Repeat last step at increased size if success.
# Try random step at decreased size if failure.
# 
# SGD gets to 0.36 in 1 epoch.
# This gets to around 0.60 sloss in 1000 epochs.
# using step=0.01 grow=1.001
# There is a long period where model is stuck around 0.70
#
# It turns out the grow factor is very important
# grow  sloss@64epoch
# 1.001	0.91
# 1.01	0.61
# 1.1	0.96 (though very high variance)
#
# (epoch,sloss,stepsize) using step=0.01 grow=1.01
# (epochs,step,grow) = (1024,0.01,1.01)
# (epochs,step,grow) = (1024,0.01,1.01)
# (1,1.7985317f0,0.0051531769028983964)
# (2,1.3769107f0,0.002341992948164361)
# (4,1.0358456f0,0.0011349164150302124)
# (8,0.8762574f0,0.00146262243400234)
# (16,0.7483307f0,0.0008924781269490016)
# (32,0.6375183f0,0.0008380797458788005)
# (64,0.5952995f0,0.0007391564809047644)
# (128,0.5699461f0,0.0003445230915309831)
# (256,0.59447426f0,0.0005493959444114676)
# (512,0.57121503f0,0.0002871048146276168)
# (1024,0.6157344f0,0.0008438423214864373)


function dhc01(; epochs=64, step=0.01, grow=1.01)
    @show (epochs,step,grow)
    global w = fill!(CudaArray(Float32,794*32),0)
    global f = initmodel(w)
    global dw = randn!(similar(w))
    global w0 = copy(w)
    nextn = 1
    for epoch in 1:epochs
        sloss = iter = ssize = 0
        for (x,y) in mnist
            f0 = softloss(forw(f,x),y)
            f1 = 0
            while true
                axpy!(step, dw, w)
                f1 = softloss(forw(f,x),y)
                if f1 < f0
                    sloss += f1; ssize += step; iter += 1
                    copy!(w0, w)
                    step *= grow
                    break
                else
                    copy!(w, w0)
                    step > 1e-6 && (step /= grow)
                    randn!(dw)
                end
            end
        end
        if epoch == nextn || epoch == epochs
            println((epoch,sloss/iter,ssize/iter))
            nextn*=2
        end
    end
end

# DHC1 is basically Procedure 3.1 from AITR1569.
# Let's try the original DHC algorithm, Procedure 4.1.

function dhc02(; epochs=64, step=0.01, grow=1.01, maxiter=5)
    @show (epochs,step,grow)
    global w = fill!(CudaArray(Float32,794*32),0)
    global w0 = copy(w)
    global u = fill!(similar(w), 0)
    global v = randn!(similar(w))
    global f = initmodel(w)
    nextn = 1
    for epoch in 1:epochs
        sloss = ssize = steps = 0
        for (x,y) in mnist
            f0 = softloss(forw(f,x),y)
            axpy!(step, v, w)
            f1 = softloss(forw(f,x),y)
            iter = 0
            while f1 > f0 && iter < maxiter
                iter += 1
                copy!(w, w0)
                randn!(v)
                axpy!(step, v, w)
                f1 = softloss(forw(f,x),y)
            end
            sloss += f1
            ssize += step
            steps += 1
            if f1 > f0
                step /= grow
            elseif iter == 0
                copy!(w0, w)
                axpy!(step, v, u)
                step *= grow
            else
                copy!(w0, w)    # save w+v
                axpy!(1, u, w)  # try w+v+u
                f2 = softloss(forw(f,x),y)
                if f2 > f1      # u+v did not work
                    copy!(w, w0)
                    fill!(u, 0) # reset memory
                    axpy!(step, v, u) # u is scaled v is not
                    step *= grow
                else
                    copy!(w0, w)
                    axpy!(step, v, u)
                    step *= grow
                end
            end
        end
        if epoch == nextn || epoch == epochs
            println((epoch,sloss/steps,ssize/steps))
            nextn*=2
        end
    end
end


# function train02(; epochs=1000, lr=0.5, adam=false, batchsize=100, seed=0, target=nothing, o...)
#     seed > 0 && setseed(seed)
#     global f = compile(:model02; o...)
#     global d = minibatch(MNIST.xtrn, MNIST.ytrn, batchsize)

#     # algorithm parameters
#     nhistory = 10                # remember last history steps
#     ihistory = 0
#     stepsize = 0                # dx = stepsize*gradient + (minstep + rndstep*stepsize)*random
#     minstep = 1e-4
#     rndstep = 0.1
    
#     for epoch=1:epochs
#         zloss = sloss = iter = 0
#         for (x,ygold) in d
#             ypred = forw(f, x)
#             sl = softloss(ypred, ygold)
#             zl = zeroone(ypred, ygold)
#             iter += 1; zloss += zl; sloss += sl

            

#         end
#         println((epoch, sloss/iter, zloss/iter))
#         zloss == 0 && break
#     end
#     return f
# end

# weights(f::Net)=map(r->r.out, params(f))
# gradients(f::Net)=map(r->r.dif, params(f))

# function vnorm(w)
#     sqnorm = 0
#     for u in w; sqnorm += vecnorm(u)^2; end
#     sqrt(sqnorm)
# end

# function vcos(u,v)
#     uv = 0
#     for i=1:length(u); uv += CUBLAS.dot(u[i],v[i]); end
#     uv / (vnorm(u)*vnorm(v))
# end

# function test02(f, data, loss)
#     sumloss = numloss = 0
#     for (x,ygold) in data
#         ypred = forw(f, x)
#         sumloss += loss(ypred, ygold)
#         numloss += 1
#     end
#     sumloss / numloss
# end

# f1 = train02(seed=1)
# f2 = train02(seed=1, target=f1)

# Can get to zloss=0 with hidden=24 using adam=true lr=0.001

#     setp(f, lr=lr, adam=adam)
            # # back(f, ygold, softloss)

            # if target != nothing
            #     # we start at w0 and want to reach w2
            #     # currently we are at w1 and about to take the step dw
            #     # compare dw and w2-w1
            #     # compare w2-w1 and w2-w0
            #     w1 = weights(f)
            #     dw = gradients(f)
            #     if w0 == nothing
            #         w0 = deepcopy(w1)
            #         w2 = weights(target)
            #         w2_w0 = deepcopy(w2)
            #         w2_w1 = deepcopy(w2)
            #         for i=1:length(w2); axpy!(-1,w0[i],w2_w0[i]); end
            #     end
            #     for i=1:length(w2); copy!(w2_w1[i],w2[i]); axpy!(-1,w1[i],w2_w1[i]); end
            #     @printf("w2-w0=%g,1 w2-w1=%g,%g dw=%g,%g\n",
            #             vnorm(w2_w0), vnorm(w2_w1), vcos(w2_w1, w2_w0),
            #             lr*vnorm(dw), -vcos(dw, w2_w1))
            # end

            # # update!(f)

# This doesn't work because initforw copies the arrays.
# @knet function model02(x; ftype=Float32, xinput=784, hidden=32, output=10,
#                        wvec=fill!(CudaArray(ftype,xinput*hidden+hidden*output),0))
#     h    = wf(x; f=:relu, winit=CudaArray(wvec.ptr, (hidden,xinput), wvec.dev))
#     return wf(h; f=:soft, winit=CudaArray(wvec.ptr+hidden*xinput*sizeof(ftype), (output,hidden), wvec.dev))
# end
